compiler/rustc_mir_transform/src/early_otherwise_branch.rs

```diff
@@ -315,89 +273,166 @@ fn evaluate_candidate<'tcx>(
         return None;
     };
     let parent_ty = parent_discr.ty(body.local_decls(), tcx);
-    let parent_dest = {
-        let poss = targets.otherwise();
-        // If the fallthrough on the parent is trivially unreachable, we can let the
-        // children choose the destination
-        if bbs[poss].statements.len() == 0
-            && bbs[poss].terminator().kind == TerminatorKind::Unreachable
-        {
-            None
-        } else {
-            Some(poss)
-        }
-    };
-    let (_, child) = targets.iter().next()?;
-    let child_terminator = &bbs[child].terminator();
-    let TerminatorKind::SwitchInt { targets: child_targets, discr: child_discr } =
-        &child_terminator.kind
+    let mut targets_iter = targets.iter();
+    let (_, first_child) = targets_iter.next()?;
+    let first_child_terminator = &bbs[first_child].terminator();
+    let TerminatorKind::SwitchInt { targets: first_child_targets, discr: first_child_discr } =
+        &first_child_terminator.kind
     else {
         return None;
     };
-    let child_ty = child_discr.ty(body.local_decls(), tcx);
-    if child_ty != parent_ty {
-        return None;
-    }
-    let Some(StatementKind::Assign(boxed)) = &bbs[child].statements.first().map(|x| &x.kind) else {
+    let hoist_discriminant = if bbs[first_child].statements.len() == 1 {
+        if !bbs[targets.otherwise()].is_empty_unreachable() {
+            // Someone could write code like this:
+            // ```rust
+            // let Q = val;
+            // if discriminant(P) == otherwise {
+            //     let ptr = &mut Q as *mut _ as *mut u8;
+            //     // Any invalid value for the type. It is possible to be opaque, such as in other functions.
+            //     unsafe { *ptr = 10; }
+            // }
+            //
+            // match P {
+            //    A => match Q {
+            //        A => {
+            //            // code
+            //        }
+            //        _ => {
+            //            // don't use Q
+            //        }
+            //    }
+            //    _ => {
+            //        // don't use Q
+            //    }
+            // };
+            // ```
+            //
+            // Hoisting the `discriminant(Q)` out of the `A` arm causes us to compute the discriminant of an
+            // invalid value, which is UB.
+            // In order to fix this, we would either need to show that the discriminant computation of
+            // `place` is computed in all branches.
+            // So we need the `otherwise` branch has no statements and an unreachable terminator.
```

Comment:
The opt would be correct if we compute the discriminant in the otherwise branch like all other ones, wouldn't it?

Reply:
Yes. But I think it could make this pass become complicated. So I decided to keep it simple first. This way, at least the pass can start to do work, and iterate over it in follow-on updates.